### PR TITLE
Improvements to CMake Build Structure

### PR DESCRIPTION
**Pull request by ambasta**
Hi @lperron 

I've tried to address some of the issues with these patches.

For one, there are sane defaults set to CMAKE_INSTALL_<DIRECTORIES> for non-Linux platforms.

Second, I've enabled fallback for gflags, glog and protobuf for systems where the aforesaid aren't already installed.

Please don't merge this just yet. I'm still making further changes and testing it out for it to be cross-platform complete

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -14,19 +15,56 @@ IF(UNIX)
...
+
+SET(${PROJECT_NAME}externalLibraries "")
+SET(${PROJECT_NAME}externalTargets "")
 
 IF(BUILD_CXX)
-    SET(CMAKE_CXX_FLAGS  "${CMAKE_CXX_FLAGS} -DUSE_GLOP -DUSE_BOP")
+    FIND_PACKAGE(Threads REQUIRED)
 
-    FIND_PACKAGE(Protobuf REQUIRED)
+    SET(Protobuf_VERSION 3.0.2)
```
Comment (lperron) [inline CMakeLists.txt]: 3.0.2, I believe the latest version is 3.3.0

Comment (ambasta) [inline CMakeLists.txt]: It is! However, I've kept it the same as the one in dependencies folder. Ideal would be to pick all of these up from a command config file that both current make and CMake can use and you could bump at will!

File: cmake/external/zlib.cmake [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+SET(Zlib_INSTALL ${CMAKE_CURRENT_BINARY_DIR}/zlib/install)
+SET(Zlib_TAG 50893291621658f355bc5b4d450a8d06a563053d)
+
+IF(WIN32)
+    SET(Zlib_STATIC_LIBRARIES
+            debug ${CMAKE_CURRENT_BINARY_DIR}/zlib/install/lib/zlibstaticd.lib
+            optimized ${CMAKE_CURRENT_BINARY_DIR}/zlib/install/lib/zlibstatic.lib)
+ELSE()
+    SET(Zlib_STATIC_LIBRARIES
+            ${CMAKE_CURRENT_BINARY_DIR}/zlib/install/lib/libz.a)
```
Comment (lperron) [inline cmake/external/zlib.cmake]: libz is always there on unix platforms. Do you want an extra flag?

**Comment by lperron** [comment]
dependency is 3.2.0, please bump to 3.3.0.

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-31 15:46 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

**Comment by lperron** [comment]
Do you want me to merge?

This looks so nice.

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-31 15:49 GMT+02:00 Laurent Perron <lperron@google.com>:

File: cmake/external/zlib.cmake (hunk 2 above)
Comment (ambasta) [inline cmake/external/zlib.cmake]: Yes, but not necessarily on windows

**Comment by ambasta** [comment]
Not just yet. Please give me a little more time to improve on some of the missing deps

**Comment by lperron** [comment]
I was commenting the non WIN32 part.


Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-31 15:51 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

File: cmake/external/zlib.cmake (hunk 2 above)
Comment (ambasta) [inline cmake/external/zlib.cmake]: Oh right, my bad! Will rectify

**Comment by lperron** [comment]
As you may have seen, I have simplified the make third_party installation
process.
It now installs gflags, glog, protobuf with cmake, and cbc with the system
installed configure.
I removed everything else.



Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-05-31 15:57 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

**Comment by ambasta** [comment]
Hi @lperron 
I've moved the whole versioning bit to a separate config file all together which both CMake and any other build systems should be able to use!

Additionally, Win build did not work so reworking that. Also forced Zlib as a system dependency.

**Comment by lperron** [comment]
-- The CXX compiler identification is AppleClang 8.1.0.8020042
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.2") 
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Checking for module 'libglog>=0.3.5'
--   Found libglog, version 0.3.5
-- Checking for module 'cbc'
--   Found cbc, version 2.9.6
CMake Error at CMakeLists.txt:79 (ADD_DEPENDENCIES):
  add_dependencies called with incorrect number of arguments

PROJECT_NAME seems undefined

**Comment by ambasta** [comment]
Hi @lperron 

I am not sure why your build is failing. Would it be possible to attach a full build log? I've been able to partially build or-tools via cmake on OSX myself.

Please find the build log here! https://gist.github.com/anonymous/dabb84cbee9ff05c25a25b9f06dcb391

**Comment by ambasta** [comment]
Acutally, with another small change part of PR #404 , builds succeed on OSx as well! https://gist.github.com/anonymous/4296aa613521c5847f1144aa57e83367

**Comment by ambasta** [comment]
Hi @lperron 

Could you please verify if you're able to build successfully? You might want to brew install pkg-config for it to work correctly.

Also, what next would you like to see in the CMake build?

**Comment by lperron** [comment]
It does not, incorrect syntax line 59 on mac (see my log on the pull
request).
I do have brew installed (with some packages on it).

Do you know if I can force cmake to recompile packages locally?

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-06-06 11:02 GMT+02:00 Amit Prakash Ambasta <notifications@github.com>:

**Comment by lperron** [comment]
I am using cmake 3.8.1 btw.

Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00


2017-06-06 11:18 GMT+02:00 Laurent Perron <lperron@google.com>:

**Pull request merged**